src/graph/Graph.java

```diff
@@ -0,0 +1,172 @@
+
+import java.util.*;
+
+public class Graph {
+
+  private Map<Integer, Vertex> vertices;
+  private String name;
+
+  public Graph() {
+    this("unnamed");
+  }
+
+  public Graph(String name) {
+    this.vertices = new HashMap<>();
+    this.name = name;
+  }
+
+  public void setName(String name) {
+    this.name = name;
+  }
+
+  public String getName() {
+    return name;
+  }
+
+  public Vertex getNode(int id) {
+    return this.vertices.get(id);
+  }
+
+  public Set<Integer> getNodesID() {
+    return vertices.keySet();
+  }
+
+  public Collection<Vertex> getNodes() {
+    return vertices.values();
+  }
+
+  public Vertex addNode(Vertex v) {
+    this.vertices.put(v.getId(), v);
+    return v;
+  }
+
+  public Vertex addNode(int id) {
+    Vertex node = new Vertex(id);
+    return this.addNode(node);
+  }
+
+  public Vertex addNode() {
+    if (this.isEmpty()) {
+      return this.addNode(0);
+    }
+    int max_id = Collections.max(this.vertices.keySet());
+    return this.addNode(max_id + 1);
+  }
+
+  public void removeNode(int id) {
+    for (Vertex v : this.getNodes()) {
+      v.removeNeighbor(id);
+    }
+    this.vertices.remove(id);
+  }
+
+  public void removeNode(Vertex v) {
+    this.removeNode(v.getId());
+  }
+
+  public Edge addEdge(Vertex node_u, Vertex node_v) {
+    node_v.addNeighbor(node_u);
+    return node_u.addNeighbor(node_v);
+  }
+
+  public Edge addEdge(int u, int v) {
+    Vertex node_u = this.vertices.get(u);
+    Vertex node_v = this.vertices.get(v);
+    if (node_u == null) {
+      node_u = this.addNode(u);
+    }
+    if (node_v == null) {
+      node_v = this.addNode(v);
+    }
+    return this.addEdge(node_u, node_v);
+  }
+
+  public void removeEdge(int u, int v) {
+    Vertex node_u = this.getNode(u);
+    Vertex node_v = this.getNode(v);
+    node_u.removeNeighbor(v);
+    node_v.removeNeighbor(u);
+  }
+
+  public void removeEdge(Vertex node_u, Vertex node_v) {
+    this.removeEdge(node_u.getId(), node_v.getId());
+  }
+
+  public void removeEdge(Edge e) {
+    this.removeEdge(e.getStart(), e.getEnd());
+  }
+
+  public List<Edge> getEdges() {
+    List<Edge> result = new ArrayList<>();
+    for (Vertex v : this.getNodes()) {
+      for (Edge e : v.getEdges()) {
+        if (e.getStart().getId() < e.getEnd().getId()) {
+          result.add(e);
+        }
+      }
+    }
+    return result;
+  }
+
+  public Map<Integer, Integer> getDegrees() {
+    Map<Integer, Integer> result = new HashMap<>();
+    for (Vertex n : this.getNodes()) {
+      result.put(n.getId(), n.getDegree());
+    }
+    return result;
+  }
+
+  public int getDegree(int id) {
+    return this.getNode(id).getDegree();
+  }
+
+  public int getDegree(Vertex v) {
+    return this.getDegree(v.getId());
+  }
+
+  public int size() {
+    return this.vertices.size();
+  }
+
+  public int getNumOfEdges() {
+    return this.getEdges().size();
+  }
+
+  public boolean isEmpty() {
+    return this.size() == 0;
+  }
+
+  public void empty() {
```

Comment:
clear() may be a better name. If you want to make clear solid, better clear the adjacency list in each vertex as well, that will help gabbage collector collect the gabbage. 
for each verex in vertices:
    vertex.edges.clear()
    vertex.edges = null
vertices.clear()